File: couch-server/src/main/java/com/fronde/server/services/validation/rule/DatePastRule.java
package com.fronde.server.services.validation.rule;

import com.fronde.server.services.validation.ValidationResultFactory;
import java.util.Date;

public class DatePastRule extends AbstractRule {

  @Override
  public void validate(Object object, Object propertyValue, ValidationResultFactory resultFactory) {
    if (propertyValue != null) {
      if (propertyValue instanceof Date) {
        Date dateValue = (Date) propertyValue;
        if (!dateValue.before(new Date())) {
          resultFactory.addResult("DatePast");
        }
      } else {
        throw new RuntimeException(
            "This rule class is expecting a Date but was given " + propertyValue.getClass()
                .getName());
      }
    }
  }
}

File: angular-client/src/app/form/validators/date-past-rule.ts
import { AbstractControl, ValidationErrors } from '@angular/forms';
import * as moment from 'moment';
import { AbstractRule } from './abstract-rule';
import { ValidationResultFactory } from './validation-result-factory';

export class DatePastRule extends AbstractRule {
  validate(control: AbstractControl): ValidationErrors | null {
    const resultFactory = new ValidationResultFactory(
      'DatePastRule',
      this.validationService
    );

    if (control.value) {
      if (!moment(control.value).isBefore(moment())) {
        resultFactory.addResult('DatePast');
      }
    }

    return resultFactory.validationErrors;
  }
}
